### PR TITLE
[IMP] web: list: select range of records with shift

### PR DESCRIPTION
**Pull request by fdardenne**
Before, in the list view, the user could not select a record range easily. He had to
check/uncheck manually each checkbox in the range.

Now, the user can select a range of records in list view by hitting its first record
and and its last record while holding SHIFT for the second.

It's also possible to do it with the keyboard:
- pressing < SPACE > to select a record and <SHIFT+SPACE> to select a range of records.
- pressing <SHIFT+ARROW> to extend to current checkbox

shiftList

task-2039894

### CODE REVIEW

File: addons/web/static/src/core/checkbox/checkbox.js [hunk 1]
```diff
@@ -48,13 +49,14 @@ export class CheckBox extends Component {
...
-            this.props.onChange(input.checked);
+            this.props.onChange(input.checked, ev.shiftKey);
         }
     }
 
     onChange(ev) {
         if (!this.props.disabled) {
-            this.props.onChange(ev.target.checked);
+            this.props.onChange(ev.target.checked, this.shiftKey);
+            this.shiftKey = false;
```
Comment (brboi) [inline addons/web/static/src/core/checkbox/checkbox.js:59]: the whole diff in this file should be removed IMHO

Instead you should make the list_renderer aware of the shiftKey status.

For instance with something similar to:
```js
useExternalListener(window, "keydown", (ev)=> { this.shiftKeyMode = ev.shiftKey; });
useExternalListener(window, "keyup", ()=> { this.shiftKeyMode = false; });
```

File: addons/web/static/src/views/list/list_renderer.js [hunk 2]
```diff
@@ -1387,6 +1392,50 @@ export class ListRenderer extends Component {
...
+                if (isExpanding) {
+                    nextRecord.toggleSelection(this.firstArrowShiftRecord.record.selected);
+                } else {
+                    record.toggleSelection(false);
+                }
+                this.lastCheckedRecord = nextRecord;
+                toFocus = this.findFocusFutureCell(cell, cellIsInGroupRow, "up");
+                break;
+            }
+            case "shift+ ":
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:1432]: this should not work:
- if it is there's an issue to fix in the hotkey service :-)
- if it isn't you did not test this use case ^^

the correct hotkey string should be "shift+space"

File: addons/web/static/src/views/list/list_renderer.js [hunk 3]
```diff
@@ -996,11 +996,9 @@ export class ListRenderer extends Component {
         }
 
         const hotkey = getActiveHotkey(ev);
-
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:999]: nitpicking: unnecessary diff

File: addons/web/static/src/views/list/list_renderer.js [hunk 4]
```diff
@@ -996,11 +996,9 @@ export class ListRenderer extends Component {
...
         if (ev.target.tagName === "TEXTAREA" && hotkey === "enter") {
             return;
         }
-
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:1003]: nitpicking: unnecessary diff

File: addons/web/static/src/views/list/list_renderer.js [hunk 5]
```diff
@@ -1015,6 +1013,13 @@ export class ListRenderer extends Component {
         }
     }
 
+    onCellKeyup(ev) {
+        const hotkey = getActiveHotkey(ev);
+        if (hotkey === "shift") {
+            this.firstArrowShiftRecord = undefined;
+        }
+    }
+
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:1025]: if you do what I suggested in the comment for the checkbox file, this should be moved within the same listener

File: addons/web/static/src/views/list/list_renderer.js [hunk 6]
```diff
@@ -1387,6 +1392,50 @@ export class ListRenderer extends Component {
                     toFocus = buttons[index - 1] || currentButton;
                 }
                 break;
+            case "shift+arrowdown": {
+                const recordIndex = this.props.list.records.indexOf(record);
+                if (recordIndex >= this.props.list.records.length - 1) {
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:1400]: IMO it would be clearer with a strictEqual operator ===

File: addons/web/static/src/views/list/list_renderer.js [hunk 7]
```diff
@@ -1517,8 +1566,21 @@ export class ListRenderer extends Component {
...
+            const end = Math.max(recordIndex, lastCheckedRecordIndex);
+            const isSelectedTarget = record.selected;
+
+            for (let i = start; i <= end; i++) {
+                this.props.list.records[i].toggleSelection(!isSelectedTarget);
+            }
+        } else {
+            record.toggleSelection();
+        }
+        this.lastCheckedRecord = record;
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:1586]: I would like to see this factorized and also for stuff in cellKeydownReadonlyMode.
Not sure how to do it though, we could discuss it IRL ;-)

File: addons/web/static/src/views/list/list_renderer.js [hunk 8]
```diff
@@ -1387,6 +1392,50 @@ export class ListRenderer extends Component {
...
+                if (isExpanding) {
+                    nextRecord.toggleSelection(this.firstArrowShiftRecord.record.selected);
+                } else {
+                    record.toggleSelection(false);
+                }
+                this.lastCheckedRecord = nextRecord;
+                toFocus = this.findFocusFutureCell(cell, cellIsInGroupRow, "down");
+                break;
+            }
+            case "shift+arrowup": {
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:1415]: What about "shift+pageup" and "shift+pagedown" ?

This would be nice to have as well ;-)

File: addons/web/static/src/views/list/list_renderer.xml [hunk 9]
```diff
@@ -15,7 +15,7 @@
             <table t-if="showTable" t-attf-class="o_list_table table table-sm table-hover position-relative mb-0  {{props.list.isGrouped ? 'o_list_table_grouped' : 'o_list_table_ungrouped table-striped'}}" t-ref="table">
                 <thead>
                     <tr>
-                        <th t-if="hasSelectors" class="o_list_record_selector o_list_controller align-middle pe-1 cursor-pointer" tabindex="-1" t-on-keydown.synthetic="(ev) => this.onCellKeydown(ev)" t-on-click.stop="toggleSelection">
+                        <th t-if="hasSelectors" class="o_list_record_selector o_list_controller align-middle pe-1 cursor-pointer" tabindex="-1" t-on-keydown="(ev) => this.onCellKeydown(ev)" t-on-click.stop="toggleSelection">
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.xml:18]: I don't understand why this was needed

File: addons/web/static/src/views/list/list_renderer.xml [hunk 10]
```diff
@@ -217,8 +217,8 @@
             t-on-touchend="() => this.onRowTouchEnd(record)"
             t-on-touchmove="() => this.onRowTouchMove(record)"
         >
-            <td t-on-keydown.synthetic="(ev) => this.onCellKeydown(ev, group, record)" t-if="hasSelectors" class="o_list_record_selector" t-on-click.stop="() => this.toggleRecordSelection(record)" tabindex="-1">
-                <CheckBox disabled="!!props.list.editedRecord or props.list.model.useSampleModel" value="record.selected" onChange.bind="() => this.toggleRecordSelection(record)" />
+            <td t-on-keydown.synthetic="(ev) => this.onCellKeydown(ev, group, record)" t-on-keyup="(ev) => this.onCellKeyup(ev, record)" t-if="hasSelectors" class="o_list_record_selector user-select-none" t-on-click.stop="(ev) => this.toggleRecordSelection(record, ev.shiftKey)" tabindex="-1">
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.xml:220]: why "user-select-none" is added here ?

File: addons/web/static/tests/helpers/utils.js [hunk 11]
```diff
@@ -364,12 +364,12 @@ export async function triggerScroll(
     await nextTick();
 }
 
-export function click(el, selector, skipVisibilityCheck = false) {
+export function click(el, selector, skipVisibilityCheck = false, options = {}) {
```
Comment (brboi) [inline addons/web/static/tests/helpers/utils.js:367]: should not be needed either (I am referring to the change I suggested in order to keep the shiftkeymode state in the list renderer, with keydown/up events)

**Review by brboi — COMMENTED** [review]
Good job ;-) Really cool feature
The list renderer keynav is not an easy topic.

File: addons/web/static/src/views/list/list_renderer.xml (hunk 10 above)
Comment (fdardenne) [reply addons/web/static/src/views/list/list_renderer.xml:220]: If we clicked in a checkbox and then shfit+click in a td below was causing the selection of all the text above. We don't want to select anything when clicking on a td, therefore the `user-select-none`

File: addons/web/static/src/views/list/list_renderer.js (hunk 8 above)
Comment (fdardenne) [reply addons/web/static/src/views/list/list_renderer.js:1435]: It's true that it could be a nice feature 👍 . I think it should be specified and be part of a new task

File: addons/web/static/src/views/list/list_renderer.js [hunk 12]
```diff
@@ -1387,6 +1427,26 @@ export class ListRenderer extends Component {
...
+            }
+            case "shift+space":
+                this.toggleRecordSelection(record);
+                toFocus = cell;
+                break;
+            case "shift":
+                this.firstArrowShiftRecord = {
+                    record,
+                    index: this.props.list.records.indexOf(record),
+                };
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:1453]: I find the name `shiftKeyedRecord` more appropriate for these reasons:
- not semantically bound to arrows
- I think it is misleading to call it "first" or "last" or something as it is just "the record on which we pressed the shift key and only that key"
- it it shorter :-)

Also, I don't like the object here. The name suggest this is a record but this is not. Why not compute the index where it is needed and simply store the record here ?
```suggestion
this.shiftKeyedRecord = record;
```

File: addons/web/static/src/views/list/list_renderer.js [hunk 13]
```diff
@@ -1059,6 +1073,34 @@ export class ListRenderer extends Component {
...
     }
 
+    expandCheckboxes(record, direction) {
+        const recordIndex = this.props.list.records.indexOf(record);
+        let nextRecord;
+        let isExpanding;
+        switch (direction) {
+            case "up":
+                if (recordIndex <= 0) {
+                    return false;
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:1086]: you never use this return value

File: addons/web/static/src/views/list/list_renderer.js [hunk 14]
```diff
@@ -1059,6 +1073,34 @@ export class ListRenderer extends Component {
...
+                }
+                nextRecord = this.props.list.records[recordIndex - 1];
+                isExpanding = this.firstArrowShiftRecord.index > recordIndex - 1;
+                break;
+            case "down":
+                if (recordIndex === this.props.list.records.length - 1) {
+                    return false;
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:1093]: same here

File: addons/web/static/src/views/list/list_renderer.js [hunk 15]
```diff
@@ -1059,6 +1073,34 @@ export class ListRenderer extends Component {
         return null;
     }
 
+    expandCheckboxes(record, direction) {
+        const recordIndex = this.props.list.records.indexOf(record);
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:1080]: This comment applies to the whole diff you made in this file.

nitpicking: IMO when you access same value more than once in nested objects, it is more readable to destructure. Furthermore, It is a micro optimization.
```suggestion
const { records } = this.props.list;
const recordIndex = records.indexOf(record);
```
then replace `this.props.list.records` below with just `records`

File: addons/web/static/src/views/list/list_renderer.js (hunk 8 above)
Comment (brboi) [reply addons/web/static/src/views/list/list_renderer.js:1440]: @bouvyd what do you think ?

File: addons/web/static/tests/views/list_view_tests.js [hunk 16]
```diff
@@ -1993,8 +2126,7 @@ QUnit.module("Views", (hooks) => {
                 type: "list",
                 resModel: "foo",
                 serverData,
-                arch:
-                    '<tree js_class="custom_list" editable="top"><field name="foo" required="1"/></tree>',
+                arch: '<tree js_class="custom_list" editable="top"><field name="foo" required="1"/></tree>',
```
Comment (brboi) [inline addons/web/static/tests/views/list_view_tests.js:2129]: automatic linting ? If yes that's 💯 okay

File: addons/web/static/tests/views/list_view_tests.js [hunk 17]
```diff
@@ -259,6 +259,139 @@ QUnit.module("Views", (hooks) => {
...
+    QUnit.test("select record range with shift click", async function (assert) {
+        await makeView({
+            type: "list",
+            resModel: "foo",
+            serverData,
+            arch: '<tree><field name="foo"/><field name="int_field"/></tree>',
+        });
+        await click(target.querySelectorAll(".o_data_row .o_list_record_selector input")[0]);
+        assert.containsOnce(target.querySelector(".o_cp_buttons"), ".o_list_selection_box");
+        assert.containsNone(target.querySelector(".o_list_selection_box"), ".o_list_select_domain");
```
Comment (brboi) [inline addons/web/static/tests/views/list_view_tests.js:271]: nitpicking: try to avoid unnecessary `.querySelector` thing
```suggestion
        assert.containsOnce(target, ".o_cp_buttons .o_list_selection_box");
        assert.containsNone(target, ".o_list_selection_box .o_list_select_domain");
```

File: addons/web/static/tests/views/list_view_tests.js [hunk 18]
```diff
@@ -259,6 +259,139 @@ QUnit.module("Views", (hooks) => {
...
     });
 
+    QUnit.test("select record range with shift click", async function (assert) {
+        await makeView({
+            type: "list",
+            resModel: "foo",
+            serverData,
+            arch: '<tree><field name="foo"/><field name="int_field"/></tree>',
+        });
+        await click(target.querySelectorAll(".o_data_row .o_list_record_selector input")[0]);
```
Comment (brboi) [inline addons/web/static/tests/views/list_view_tests.js:269]: even if it seems more complicated here, same nitpicking comment applies: you could probably do something like
```suggestion
        await click(target, ".o_data_row[data-id=datapoint_42] .o_list_record_selector input");
```

File: addons/web/static/src/views/list/list_renderer.js [hunk 19]
```diff
@@ -405,7 +405,10 @@ export class ListRenderer extends Component {
...
     focus(el) {
         el.focus();
-        if (["INPUT", "TEXTAREA"].includes(el.tagName)) {
+        if (
+            ["INPUT", "TEXTAREA"].includes(el.tagName) &&
+            ["text", "search", "URL", "tel", "password", "textarea"].includes(el.type)
+        ) {
             if (el.selectionStart === null) {
                 return;
             }
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:414]: Your fix makes the first nested `if (condition) {}` obsolete, isn't it ?

Also, this is not the first place in odoo where we need the same fix (e.g. look for useNumpadDecimal).
Maybe we should have an utility function i.e. `hasSelectionFeature(element)` ?

You don't have to do it in this PR, though (but you could as this is the master branch niark niark)

File: addons/web/static/src/views/list/list_renderer.js [hunk 20]
```diff
@@ -405,7 +405,10 @@ export class ListRenderer extends Component {
 
     focus(el) {
         el.focus();
-        if (["INPUT", "TEXTAREA"].includes(el.tagName)) {
+        if (
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:408]: Also... :-D
image

**Review by brboi — COMMENTED** [review]
really cool feature, I love it

File: addons/web/static/src/views/list/list_renderer.js (hunk 20 above)
Comment (brboi) [reply addons/web/static/src/views/list/list_renderer.js:408]: Hmmm I did not think it through, you're right, I do not see a way to test this, sorry I just wanted to use this meme so bad ^^

File: addons/web/static/src/views/list/list_renderer.js (hunk 12 above)
Comment (fdardenne) [reply addons/web/static/src/views/list/list_renderer.js:1453]: Agree on the name 👍 
The reason I store the index of the record is that I don't want it to be computed every time the user expands its selection with shift+arrow. Depending on `limit` imposed in the arch or the pager limit of the list view it can be long every time the user expands

File: addons/web/static/src/views/list/list_renderer.js [hunk 21]
```diff
@@ -1387,6 +1433,25 @@ export class ListRenderer extends Component {
...
+            }
+            case "shift+arrowup": {
+                if (this.expandCheckboxes(record, "up")) {
+                    toFocus = this.findFocusFutureCell(cell, cellIsInGroupRow, "up");
+                }
+                break;
+            }
+            case "shift+space":
+                this.toggleRecordSelection(record);
+                toFocus = cell;
```
Comment (brboi) [inline addons/web/static/src/views/list/list_renderer.js:1450]: toFocus = getElementToFocus(cell)

File: addons/web/static/tests/views/list_view_tests.js [hunk 22]
```diff
@@ -259,6 +259,139 @@ QUnit.module("Views", (hooks) => {
...
+            "1 selected"
+        );
+        assert.strictEqual(
+            document.querySelectorAll(".o_data_row .o_list_record_selector input:checked").length,
+            1
+        );
+
+        // shift click the 4th record to have 0-1-2-3 toggled
+        const { keydownEvent } = triggerHotkey("shift");
+        document.activeElement.dispatchEvent(keydownEvent);
```
Comment (brboi) [inline addons/web/static/tests/views/list_view_tests.js:283]:
```suggestion
await triggerEvent(document.activeElement, "keydown", { key: "Shift", shiftKey: true })
```

File: addons/web/static/tests/views/list_view_tests.js [hunk 23]
```diff
@@ -259,6 +259,150 @@ QUnit.module("Views", (hooks) => {
...
+        await click(checkbox);
+        assert.ok(checkbox.checked);
+
+        // Go to the fourth checkbox and shift+space
+        triggerHotkey("ArrowDown");
+        triggerHotkey("ArrowDown");
+        triggerHotkey("ArrowDown");
+        checkbox = target.querySelector(".o_data_row:nth-child(4) .o_list_record_selector input");
+        assert.strictEqual(document.activeElement, checkbox);
+        assert.ok(!checkbox.checked);
```
Comment (brboi) [inline addons/web/static/tests/views/list_view_tests.js:340]: please also verify this, this seems weird that there is no `await nextTick` before the two assertions.
Maybe the assertions are already true before triggering the hotkeys ?

**Review by brboi — COMMENTED** [review]
Sorry for this fourth or fifth review :-)
Please apply the suggestions.

Also, about the tests comments: the only place where you would need to
```js
        triggerEvent(document.activeElement, null, "keydown", {
            key: "Shift",
            shiftKey: true,
        });
```
is for "shift-clicking" (you don't need to await the triggerEvent as you'll `await click` right after)

File: addons/web/static/tests/views/list_view_tests.js [hunk 24]
```diff
@@ -259,6 +259,150 @@ QUnit.module("Views", (hooks) => {
...
+        await triggerEvent(document.activeElement, null, "keydown", {
+            key: "Shift",
+            shiftKey: true,
+        });
+        triggerHotkey("shift+space");
+        await nextTick();
```
[the remark this answers is not in the document]
Comment (fdardenne) [reply addons/web/static/tests/views/list_view_tests.js:346]: Pressing the shift key before is necessary as `toggleRecordSelection` use `this.shiftKeyMode` to know if the shift mode is used or not.
Functionally, it's the same, when pressing <shift+space>, first the browser trigger <shift> and then trigger <shift+space>

File: addons/web/static/tests/views/list_view_tests.js [hunk 25]
```diff
@@ -259,6 +259,150 @@ QUnit.module("Views", (hooks) => {
...
+        );
+        assert.strictEqual(document.activeElement, checkbox);
+        await click(checkbox);
+        assert.ok(checkbox.checked);
+
+        // expand the checkbox with arrowdown
+        await triggerEvent(document.activeElement, null, "keydown", {
+            key: "Shift",
+            shiftKey: true,
+        });
```
[the remark this answers is not in the document]
Comment (fdardenne) [reply addons/web/static/tests/views/list_view_tests.js:382]: Same as above

File: addons/web/static/tests/views/list_view_tests.js (hunk 23 above)
Comment (fdardenne) [reply addons/web/static/tests/views/list_view_tests.js:340]: The `ArrowDown` are executed and are necessary for the test. The assertion is not true  by removing one or all of them.
However I put the `await nextTick` so we are sure

**Comment by brboi** [comment]
robodoo rebase-ff r+

**Comment by brboi** [comment]
robodoo rebase-ff r+
